certbot-ci/certbot_integration_tests/certbot_tests/test_main.py

```diff
@@ -19,21 +23,63 @@ def context(request):
+        '-a', 'manual', '-d', certname, '--preferred-challenges', 'dns',
+        'run', '--cert-name', certname,
+        '--manual-auth-hook', context.manual_dns_auth_hook,
+        '--manual-cleanup-hook', context.manual_dns_cleanup_hook,
+        '--pre-hook', 'echo wtf.pre >> "{0}"'.format(context.hook_probe),
+        '--post-hook', 'echo wtf.post >> "{0}"'.format(context.hook_probe),
+        '--renew-hook', 'echo renew >> "{0}"'.format(context.hook_probe)
+    ])
 
-        certname = context.domain()
-        context.certbot([
-            'certonly', '-a', 'manual', '-d', certname,
-            '--cert-name', certname,
-            '--manual-auth-hook', manual_http_hooks[0],
-            '--manual-cleanup-hook', manual_http_hooks[1],
-            '--pre-hook', 'echo wtf.pre >> "{0}"'.format(context.hook_probe),
-            '--post-hook', 'echo wtf.post >> "{0}"'.format(context.hook_probe),
-            '--deploy-hook', 'echo deploy >> "{0}"'.format(context.hook_probe)
-        ])
+    with pytest.raises(AssertionError):
+        assert_hook_execution(context.hook_probe, 'renew')
+    assert_save_renew_hook(context.config_dir, certname)
+
+
+def test_renew(context):
+    """Test various certificate renew scenarios."""
+    # First, we create a target certificate, with all hook dirs instantiated.
+    # We should have a new certificate, with hooks executed.
+    # Check also file permissions.
+    certname = context.domain('renew')
+    context.certbot([
+        'certonly', '-d', certname, '--rsa-key-size', '4096',
+        '--preferred-challenges', 'http-01'
+    ])
+
+    assert_certs_count_for_lineage(context.config_dir, certname, 1)
+    assert_world_permissions(
+        join(context.config_dir, 'archive/{0}/privkey1.pem'.format(certname)), 0)
+
+    # Second, we force renew, and ensure that renewal hooks files are executed.
+    # Also check that file permissions are correct.
+    misc.generate_test_file_hooks(context.config_dir, context.hook_probe)
+    context.certbot(['renew'])
 
+    assert_certs_count_for_lineage(context.config_dir, certname, 2)
     assert_hook_execution(context.hook_probe, 'deploy')
-    assert_save_renew_hook(context.config_dir, certname)
+    assert_world_permissions(
+        join(context.config_dir, 'archive/{0}/privkey2.pem'.format(certname)), 0)
+    assert_equals_group_owner(
+        join(context.config_dir, 'archive/{0}/privkey1.pem'.format(certname)),
+        join(context.config_dir, 'archive/{0}/privkey2.pem'.format(certname)))
+    assert_equals_permissions(
+        join(context.config_dir, 'archive/{0}/privkey1.pem'.format(certname)),
+        join(context.config_dir, 'archive/{0}/privkey2.pem'.format(certname)), 0o074)
+
+    os.chmod(join(context.config_dir, 'archive/{0}/privkey2.pem'.format(certname)), 0o444)
+
+    # Third, we try to renew without force.
+    # It is not time, so no renew should occur, and no hooks should be executed.
+    open(context.hook_probe, 'w').close()
```

Comment:
What is this line doing?

Reply:
Clean the previous content of the probe file, to assert on it correctly in the next execution of Certbot. One more argument to do the refactor of test_renew, as said below.

But please consider the current implementation in this PR as a POC to have a real call on the assertions. I will do the refactor in the next PR.

Reply:
Ah right; that would make a good comment.

I would rather see extraneous assertions removed from this PR and added in when they're needed for real code later than have a dummy POC at higher layers that'll be refactored later.

Reply:
Considering your two last comments, I reduce the test about renewal with hook scripts to what will be its final form. It still invokes all assertions.